File: propspot-os/scripts/merge-properties.py
#!/usr/bin/env python3
"""
merge-properties.py — Merges 452 Loring Dr (discard) INTO 451 Loring dr (keep).
Run via:  railway run python3 scripts/merge-properties.py
"""
import os, sys, pg8000.native

DATABASE_URL = os.environ.get('DATABASE_URL')
if not DATABASE_URL:
    print('❌ DATABASE_URL not set. Run via: railway run python3 scripts/merge-properties.py')
    sys.exit(1)

# Parse postgres://user:pass@host:port/db
from urllib.parse import urlparse
u = urlparse(DATABASE_URL)

conn = pg8000.native.Connection(
    host=u.hostname,
    port=u.port or 5432,
    database=u.path.lstrip('/'),
    user=u.username,
    password=u.password,
    ssl_context=True,
)

SIMPLE_TABLES = [
    'calendar_events','property_files','prospects','leads','opportunities',
    'purchases','projects','holdings_items','holdings_payments','holdings_documents',
    'folders','photos','share_links','work_orders','lawn_mow_events',
    'inbox_threads','inbox_attachment_saves','uw_deals','uw_audit_log','tasks',
    'drive_folders','drive_files','inkd_envelopes',
]

try:
    # Find the two properties
    rows = conn.run("SELECT id, address_line1, city, state, status FROM properties WHERE address_line1 ILIKE '%451 Loring%' LIMIT 1")
    if not rows:
        print('❌ KEEP property (451 Loring) not found'); sys.exit(1)
    keep_id, keep_addr, keep_city, keep_state, keep_status = rows[0]

    rows = conn.run("SELECT id, address_line1, city, state, status FROM properties WHERE address_line1 ILIKE '%452 Loring%' LIMIT 1")
    if not rows:
        print('❌ DISCARD property (452 Loring) not found'); sys.exit(1)
    discard_id, disc_addr, disc_city, disc_state, disc_status = rows[0]

    if keep_id == discard_id:
        print('❌ Both resolve to the same property'); sys.exit(1)

    print('\n📋 Merge plan')
    print('─' * 57)
    print(f'  KEEP    [{keep_id}]')
    print(f'          {keep_addr}, {keep_city}, {keep_state}  ({keep_status})')
    print(f'  DISCARD [{discard_id}]')
    print(f'          {disc_addr}, {disc_city}, {disc_state}  ({disc_status})')
    print('─' * 57 + '\n')

    conn.run('BEGIN')

    # Simple re-assignments
    for table in SIMPLE_TABLES:
        try:
            conn.run(f'UPDATE {table} SET property_id = :new WHERE property_id = :old',
                     new=keep_id, old=discard_id)
            n = conn.row_count
            if n > 0:
                print(f'  ✓ {table}: {n} row(s) moved')
        except pg8000.native.DatabaseError as e:
            if '42P01' in str(e):  # table not found
                print(f'  – {table}: table not found, skipping')
            else:
                raise

    # activity (entity_id)
    conn.run("UPDATE activity SET entity_id = :new WHERE entity_type = 'property' AND entity_id = :old",
             new=keep_id, old=discard_id)
    n = conn.row_count
    if n > 0: print(f'  ✓ activity: {n} row(s) moved')

    # property_contacts
    conn.run("""INSERT INTO property_contacts (property_id, contact_id, role, created_at)
                SELECT :new, contact_id, role, created_at FROM property_contacts WHERE property_id = :old
                ON CONFLICT DO NOTHING""", new=keep_id, old=discard_id)
    conn.run('DELETE FROM property_contacts WHERE property_id = :old', old=discard_id)
    if conn.row_count > 0: print('  ✓ property_contacts: merged')

    # property_access
    conn.run("""INSERT INTO property_access (property_id, user_id, granted_by, created_at)
                SELECT :new, user_id, granted_by, created_at FROM property_access WHERE property_id = :old
                ON CONFLICT DO NOTHING""", new=keep_id, old=discard_id)
    conn.run('DELETE FROM property_access WHERE property_id = :old', old=discard_id)
    if conn.row_count > 0: print('  ✓ property_access: merged')

    # lawn_maintenance (property_id is PK)
    lm_discard = conn.run('SELECT 1 FROM lawn_maintenance WHERE property_id = :id', id=discard_id)
    if lm_discard:
        lm_keep = conn.run('SELECT 1 FROM lawn_maintenance WHERE property_id = :id', id=keep_id)
        if not lm_keep:
            conn.run('UPDATE lawn_maintenance SET property_id = :new WHERE property_id = :old',
                     new=keep_id, old=discard_id)
            print('  ✓ lawn_maintenance: moved')
        else:
            conn.run('DELETE FROM lawn_maintenance WHERE property_id = :old', old=discard_id)
            print('  ✓ lawn_maintenance: discard row dropped (keep already has one)')

    # pinned_properties / recent_properties
    for table in ['pinned_properties', 'recent_properties']:
        conn.run(f"""INSERT INTO {table} (user_id, property_id)
                     SELECT user_id, :new FROM {table} WHERE property_id = :old
                     ON CONFLICT DO NOTHING""", new=keep_id, old=discard_id)
        conn.run(f'DELETE FROM {table} WHERE property_id = :old', old=discard_id)
        if conn.row_count > 0: print(f'  ✓ {table}: merged')

    # Delete the discard property
    conn.run('DELETE FROM properties WHERE id = :id', id=discard_id)

    conn.run('COMMIT')

    print('\n' + '─' * 57)
    print(f'🎉 Done!  {discard_id}')
    print(f'         ↳ all data now lives under {keep_id}')
    print('─' * 57 + '\n')

except Exception as e:
    try: conn.run('ROLLBACK')
    except: pass
    print(f'\n❌ Merge failed — database rolled back.')
    print(f'   Error: {e}')
    sys.exit(1)
finally:
    conn.close()

File: propspot-os/scripts/merge-properties.js
#!/usr/bin/env node
/**
 * merge-properties.js
 *
 * Re-homes every related record from a DISCARD property into a KEEP property,
 * then deletes the discard. Runs inside a single transaction — if anything
 * fails the database is left unchanged.
 *
 * Usage (Railway):
 *   railway run node scripts/merge-properties.js "<keep address>" "<discard address>"
 *   railway run node scripts/merge-properties.js <keep-uuid> <discard-uuid>
 *
 * Example:
 *   railway run node scripts/merge-properties.js "451 Loring" "452 Loring"
 */

require('dotenv').config();
const { Pool } = require('pg');

const pool = new Pool({ connectionString: process.env.DATABASE_URL });

// ── Tables where a plain UPDATE property_id is safe ──────────────────────────
const SIMPLE_TABLES = [
  'calendar_events',
  'property_files',
  'prospects',
  'leads',
  'opportunities',
  'purchases',
  'projects',
  'holdings_items',
  'holdings_payments',
  'holdings_documents',
  'folders',
  'photos',
  'share_links',
  'work_orders',
  'lawn_mow_events',
  'inbox_threads',
  'inbox_attachment_saves',
  'uw_deals',
  'uw_audit_log',
  'tasks',
  'drive_folders',
  'drive_files',
  'inkd_envelopes',
];

async function findProperty(client, arg) {
  // UUID?
  if (/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(arg)) {
    const { rows } = await client.query('SELECT * FROM properties WHERE id = $1', [arg]);
    return rows[0] || null;
  }
  // Address search
  const { rows } = await client.query(
    `SELECT * FROM properties WHERE address_line1 ILIKE $1 ORDER BY created_at DESC`,
    [`%${arg}%`]
  );
  if (rows.length === 0) return null;
  if (rows.length > 1) {
    console.error(`\n❌ "${arg}" matched ${rows.length} properties — be more specific or use UUID:`);
    rows.forEach(r => console.error(`   ${r.id}  ${r.address_line1}, ${r.city} (${r.status})`));
    process.exit(1);
  }
  return rows[0];
}

async function main() {
  const [,, keepArg, discardArg] = process.argv;

  if (!keepArg || !discardArg) {
    console.error('Usage: node scripts/merge-properties.js "<keep>" "<discard>"');
    console.error('       Arguments can be a UUID or an address fragment.');
    process.exit(1);
  }

  const client = await pool.connect();

  try {
    const keep    = await findProperty(client, keepArg);
    const discard = await findProperty(client, discardArg);

    if (!keep)    { console.error(`❌ Keep property not found: "${keepArg}"`);    process.exit(1); }
    if (!discard) { console.error(`❌ Discard property not found: "${discardArg}"`); process.exit(1); }
    if (keep.id === discard.id) { console.error('❌ Both arguments resolve to the same property.'); process.exit(1); }

    const NEW = keep.id;
    const OLD = discard.id;

    console.log('\n📋 Merge plan');
    console.log('─────────────────────────────────────────────────────────');
    console.log(`  KEEP    [${NEW}]`);
    console.log(`          ${keep.address_line1}, ${keep.city}, ${keep.state}  (${keep.status})`);
    console.log(`  DISCARD [${OLD}]`);
    console.log(`          ${discard.address_line1}, ${discard.city}, ${discard.state}  (${discard.status})`);
    console.log('─────────────────────────────────────────────────────────\n');

    await client.query('BEGIN');

    // ── Simple re-assignments ───────────────────────────────────────────────
    for (const table of SIMPLE_TABLES) {
      try {
        const { rowCount } = await client.query(
          `UPDATE ${table} SET property_id = $1 WHERE property_id = $2`,
          [NEW, OLD]
        );
        if (rowCount > 0) console.log(`  ✓ ${table}: ${rowCount} row(s) moved`);
      } catch (e) {
        // Table might not exist yet in older deploys — skip it
        if (e.code === '42P01') { console.log(`  – ${table}: table not found, skipping`); }
        else throw e;
      }
    }

    // ── activity — uses generic entity_id, not a property_id column ────────
    const { rowCount: actRows } = await client.query(
      `UPDATE activity SET entity_id = $1
        WHERE entity_type = 'property' AND entity_id = $2`,
      [NEW, OLD]
    );
    if (actRows > 0) console.log(`  ✓ activity: ${actRows} row(s) moved`);

    // ── property_contacts — PK (property_id, contact_id, role) ─────────────
    await client.query(`
      INSERT INTO property_contacts (property_id, contact_id, role, created_at)
        SELECT $1, contact_id, role, created_at
          FROM property_contacts WHERE property_id = $2
        ON CONFLICT DO NOTHING
    `, [NEW, OLD]);
    const { rowCount: pcDel } = await client.query(
      `DELETE FROM property_contacts WHERE property_id = $1`, [OLD]
    );
    if (pcDel > 0) console.log(`  ✓ property_contacts: merged`);

    // ── property_access — UNIQUE (property_id, user_id) ────────────────────
    await client.query(`
      INSERT INTO property_access (property_id, user_id, granted_by, created_at)
        SELECT $1, user_id, granted_by, created_at
          FROM property_access WHERE property_id = $2
        ON CONFLICT DO NOTHING
    `, [NEW, OLD]);
    const { rowCount: paDel } = await client.query(
      `DELETE FROM property_access WHERE property_id = $1`, [OLD]
    );
    if (paDel > 0) console.log(`  ✓ property_access: merged`);

    // ── lawn_maintenance — property_id IS the primary key ──────────────────
    const { rows: lmKeep }    = await client.query(
      `SELECT 1 FROM lawn_maintenance WHERE property_id = $1`, [NEW]
    );
    const { rows: lmDiscard } = await client.query(
      `SELECT 1 FROM lawn_maintenance WHERE property_id = $1`, [OLD]
    );
    if (lmDiscard.length > 0) {
      if (lmKeep.length === 0) {
        await client.query(
          `UPDATE lawn_maintenance SET property_id = $1 WHERE property_id = $2`, [NEW, OLD]
        );
        console.log(`  ✓ lawn_maintenance: moved`);
      } else {
        await client.query(`DELETE FROM lawn_maintenance WHERE property_id = $1`, [OLD]);
        console.log(`  ✓ lawn_maintenance: discard row dropped (keep already has one)`);
      }
    }

    // ── pinned_properties / recent_properties — PK (user_id, property_id) ──
    for (const table of ['pinned_properties', 'recent_properties']) {
      await client.query(`
        INSERT INTO ${table} (user_id, property_id)
          SELECT user_id, $1 FROM ${table} WHERE property_id = $2
          ON CONFLICT DO NOTHING
      `, [NEW, OLD]);
      const { rowCount: rDel } = await client.query(
        `DELETE FROM ${table} WHERE property_id = $1`, [OLD]
      );
      if (rDel > 0) console.log(`  ✓ ${table}: merged`);
    }

    // ── Finally: delete the discard property ───────────────────────────────
    await client.query(`DELETE FROM properties WHERE id = $1`, [OLD]);

    await client.query('COMMIT');

    console.log('\n─────────────────────────────────────────────────────────');
    console.log(`🎉 Done!  ${OLD}`);
    console.log(`         ↳ all data now lives under ${NEW}`);
    console.log('─────────────────────────────────────────────────────────\n');

  } catch (err) {
    await client.query('ROLLBACK').catch(() => {});
    console.error('\n❌ Merge failed — database rolled back to its original state.');
    console.error('   Error:', err.message);
    process.exit(1);
  } finally {
    client.release();
    await pool.end();
  }
}

main();
